### DISCUSSION TITLE
Send Message with IRequestReplyClient throwing Producer not started exception

### DISCUSSION THREAD

**Discussion by juanmacrod** [Havret/dotnet-activemq-artemis-client#512] [Q&A] [answered]
Hi everyone,
I'm using IRequestReplyClient to send a message to artemis in .net core 8 using the directions for DI.
However, I'm getting a Producer not started exception when I try to send a message.
I have tried registering an AnonymousProducer with DI but I'm not sure where I should be using it.

Here is an excerpt from startup:
``
        services.AddActiveMq("artemis", endpoints).AddRequestReplyClient<OTRequestReplyClient>().AddAnonymousProducer<MessageProducer>();
services.AddActiveMqHostedService();
``
In my RequestReplyClient:
`var` response = await rrClient.SendAsync(address, msg, `cancellationToken);` => throws Producer not started exception

Could someone please point me in the right direction?

Many Thanks for any help.

**Comment by Havret** [comment]
Hi @juanmacrod,

You don’t need to register additional producers to use the RequestReplyClient. The AddRequestReplyClient method is sufficient. To start the producer (and the consumer) used by the RequestReplyClient, you need to resolve the IActiveMqClient from your service provider and call StartAsync. Similarly, ensure you call StopAsync for a graceful shutdown. This is handled automatically if you’re using AddActiveMqHostedService.

If the issue persists, consider providing a minimal reproducible example or a unit test so I can help investigate further.

Thanks,
Havret

**Comment by Havret** [comment]
Hi,

Please take a look at this example:
https://havret.github.io/dotnet-activemq-artemis-client/docs/request-reply.


The key issue seems to be that you're not sending a response back to the
`ReplyTo` address. You need to acknowledge the incoming message and then
send the response to the specified `ReplyTo` address.

Everything you need is in the documentation, but if you need further
clarification, feel free to ask.

I hope this helps!
Havret

**Reply by juanmacrod** [reply]
Hi,
Many thanks for you quick response and your patience
I have tried it this way but I think I am doing something fundamentally wrong.
Here an excerpt of my code:

**Request**
**startup**
Here I initialise the RequestReplyClient
``
services.AddActiveMq("artemis",` endpoints).AddRequestReplyClient<OTRequestReplyClient>();
services.AddActiveMqHostedService();
``
**controller**
in the controller I send the message
var response = await rrClient.SendAsync(address, msg, cancellationToken);
``

**Response**
**startup**
Here I initialise the Consumer and message handler
``
services.AddActiveMq("artemis", endpoints) .AddConsumer("OT-REQUEST", ActiveMQ.Artemis.Client.RoutingType.Anycast, async (message, consumer, serviceProvider, cancelllationToken) =>
        {
            await OrderTrackingListRequestConsumer.HandleMessage("OT-REQUEST", endpoints[0], message, consumer, serviceProvider, cancelllationToken);
        });
services.AddActiveMqHostedService();
``
**message handler**
The message handler is called when there is a new Message
``
var connectionFactory = new ActiveMQ.Artemis.Client.ConnectionFactory();
var connection = await connectionFactory.CreateAsync(endpoint);
var request = await consumer.ReceiveAsync();
await using var producer = await connection.CreateAnonymousProducerAsync();
await producer.SendAsync(request.ReplyTo, new ActiveMQ.Artemis.Client.Message("test")
{
     CorrelationId = request.CorrelationId,
     TimeToLive = TimeSpan.FromSeconds(30)
});
await consumer.AcceptAsync(request);
``

Sometimes it works correctly.
What happens ca. 90% of the time, I send a message and then the consumer message handler is called but it hangs at 
`await` `consumer.ReceiveAsync();`

Any help would be greatly appreciated.
Thanks for your time.
Juan

**Comment by juanmacrod** [comment, marked as answer]
Hi,
my mistake!!!
I was calling consumer.ReceiveAsync without needing it because  the message handler already had a reference to the incoming message.
With that call removed, It's working fine now.
Many Thanks for such a great project.
